File: Heap Sort/heapSort.py
def heapify(arr: list, heapSize: int, index: int):
    largest = index
    left = index * 2 + 1
    right = index * 2 + 2

    if left < heapSize and arr[left] > arr[largest]:
        largest = left
    if right < heapSize and arr[right] > arr[largest]:
        largest = right

    if largest != index:
        arr[index], arr[largest] = arr[largest], arr[index]
        heapify(arr, heapSize, largest)

def makeHeap(arr: list, lenArr: int):
    for i in range(lenArr // 2 - 1, -1, -1):
        heapify(arr, lenArr, i)

def heapSort(arr: list):
    lenArr = len(arr)

    makeHeap(arr, lenArr)

    for i in range(lenArr - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)

arr = [1, 24, 2, 4, 1, 2]
heapSort(arr)
print(arr)

File: Heap Sort/heapSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void heapify(vector<int>& arr, int heapSize, int index) {
  int largest = index;
  int leftChild = 2 * index + 1;
  int rightChild = 2 * index + 2;

  if (leftChild < heapSize && arr[leftChild] > arr[largest]) {
    largest = leftChild;
  }
  if (rightChild < heapSize && arr[rightChild] > arr[largest]) {
    largest = rightChild;
  }

  if (largest != index) {
    swap(arr[index], arr[largest]);
    heapify(arr, heapSize, index);
  }
}

void heapifyMin(vector<int>& arr, int heapSize, int index) {
    int smallest = index;
    int leftChild = 2 * index + 1;
    int rightChild = 2 * index + 2;

  
    if (leftChild < heapSize && arr[leftChild] < arr[smallest]) {
        smallest = leftChild;
    }
    if (rightChild < heapSize && arr[rightChild] < arr[smallest]) {
        smallest = rightChild;
    }

  
    if (smallest != index) {
        swap(arr[index], arr[smallest]);
        heapifyMin(arr, heapSize, smallest);
    }
}


void buildHeap(vector<int>& arr, int len) {
  for (int i = len / 2 - 1; i >= 0; --i) {
    heapify(arr, len, i);
  }
}

void buildHeapMin(vector<int>& arr, int len) {
  for (int i = len / 2 - 1; i >= 0; --i) {
    heapifyMin(arr, len, i);
  }
}



void heapSort(vector<int>& arr) {
  int len = arr.size();

  buildHeap(arr, len);

  for (int i = len - 1; i > 0; --i) {
    swap(arr[0], arr[i]);

    heapify(arr, i, 0);
  }
}

void heapSortRev(vector<int>& arr) {
  int len = arr.size();

  buildHeapMin(arr, len);

  for (int i = len - 1; i > 0; --i) {
    swap(arr[0], arr[i]);

    heapifyMin(arr, i, 0);
  }
}



int main() {
  vector<int> arr = {4, 10, 3, 5, 1};

  heapSortRev(arr);

  cout << "Sorted array: ";
  for (int num : arr) {
    cout << num << " ";
  }
  cout << endl;

  return 0;
}